File: platform/android/MapLibreAndroidTestApp/src/androidTest/java/org/maplibre/android/testapp/annotations/PolylineTest.java
package org.maplibre.android.testapp.annotations;

import android.graphics.Color;

import org.maplibre.android.annotations.Polyline;
import org.maplibre.android.annotations.PolylineOptions;
import org.maplibre.android.geometry.LatLng;
import org.maplibre.android.testapp.activity.EspressoTest;

import org.junit.Ignore;
import org.junit.Test;

import static org.maplibre.android.testapp.action.MapLibreMapAction.invoke;
import static org.junit.Assert.assertEquals;

public class PolylineTest extends EspressoTest {

  @Test
  @Ignore
  public void addPolylineTest() {
    validateTestSetup();
    invoke(maplibreMap, (uiController, maplibreMap) -> {
      LatLng latLngOne = new LatLng();
      LatLng latLngTwo = new LatLng(1, 0);

      assertEquals("Polygons should be empty", 0, maplibreMap.getPolygons().size());

      final PolylineOptions options = new PolylineOptions();
      options.color(Color.BLUE);
      options.add(latLngOne);
      options.add(latLngTwo);
      Polyline polyline = maplibreMap.addPolyline(options);

      assertEquals("Polylines should be 1", 1, maplibreMap.getPolylines().size());
      assertEquals("Polyline id should be 0", 0, polyline.getId());
      assertEquals("Polyline points size should match", 2, polyline.getPoints().size());
      assertEquals("Polyline stroke color should match", Color.BLUE, polyline.getColor());
      maplibreMap.clear();
      assertEquals("Polyline should be empty", 0, maplibreMap.getPolylines().size());
    });
  }
}

File: platform/android/MapLibreAndroid/src/test/java/org/maplibre/android/annotations/PolylineTest.kt
package org.maplibre.android.annotations

import org.maplibre.android.geometry.LatLng
import org.junit.Assert
import org.junit.Test
import org.maplibre.android.BaseTest

class PolylineTest : BaseTest() {
    @Test
    fun testSanity() {
        val polylineOptions = PolylineOptions()
        Assert.assertNotNull("polylineOptions should not be null", polylineOptions)
    }

    @Test
    fun testPolyline() {
        val polyline = PolylineOptions().polyline
        Assert.assertNotNull("polyline should not be null", polyline)
    }

    @Test
    fun testAlpha() {
        val polyline = PolylineOptions().alpha(0.2f).polyline
        Assert.assertEquals(0.2f, polyline.alpha, 0.0f)
    }

    @Test
    fun testWidth() {
        val polyline = PolylineOptions().width(1f).polyline
        Assert.assertEquals(1.0f, polyline.width, 0f)
    }

    @Test
    fun testColor() {
        val polyline = PolylineOptions().color(1).polyline
        Assert.assertEquals(1, polyline.color.toLong())
    }

    @Test
    fun testAddLatLng() {
        val polyline = PolylineOptions().add(LatLng(0.0, 0.0)).polyline
        Assert.assertNotNull("Points should not be null", polyline.points)
        Assert.assertEquals(LatLng(0.0, 0.0), polyline.points[0])
    }

    @Test
    fun testAddAllLatLng() {
        val coordinates: MutableList<LatLng> = ArrayList()
        coordinates.add(LatLng(0.0, 0.0))
        val polyline = PolylineOptions().addAll(coordinates).polyline
        Assert.assertNotNull(polyline.points)
        Assert.assertEquals(LatLng(0.0, 0.0), polyline.points[0])
    }

    @Test
    fun testBuilder() {
        val polylineOptions = PolylineOptions()
        polylineOptions.width(1.0f)
        polylineOptions.color(2)
        polylineOptions.add(LatLng(0.0, 0.0))
        val polyline = polylineOptions.polyline
        Assert.assertEquals(1.0f, polyline.width, 0f)
        Assert.assertEquals(2, polyline.color.toLong())
        Assert.assertNotNull("Points should not be null", polyline.points)
        Assert.assertEquals(LatLng(0.0, 0.0), polyline.points[0])
    }
}
